### ISSUE TITLE
commit confusion

### ISSUE THREAD

**Issue by alvint** [TablePlus/TablePlus#818]
Oracle 12.x, build 159

When executing a statement that changes rows such as `update my_table set foo = 1 where id = 3;`, the changes aren't committed to the DB immediately--I believe this is intended behavior.  However, hitting CMD-S to commit the changes does not work. Instead I have to manually type `commit;`. Is this intended behavior?

My opinion is that allowing CMD-S to commit for this case as well would be more consistent; CMD-S would mean save your changes no matter how you made them. But I haven't spent a lot of time thinking about it so maybe there's a reason this should be done.

In any case, how about adding some visual indication to let users know that the changes haven't been flushed to the DB yet? I constantly forget to commit since most of the tools I used before had auto-commit.

**Comment by alvint** [comment]
Another suggestion is that once the changes are committed, how about automatically refreshing any table views?

**Comment by huyphams** [comment]
Hi, @alvint TablePlus has auto-commit. if you press cmd + S, it will send the "COMMIT" after sending all the update queries. However, it will not send commit if you making a change via SQL Editor.
It also refreshes all opened tables if you making a change by pressing cmd + S.

**Comment by alvint** [comment]
TablePlus does have auto-commit? How do you turn it on?

**Comment by alvint** [comment]
By auto-commit, you mean it will commit the changes when you make changes in the table view and hit cmd-S, but not when making changes in the editor?

**Comment by huyphams** [comment]
That's true @alvint You have to manual commit when making a change via the editor.

**Comment by alvint** [comment]
I still like the cmd-S after manual changes idea, and/or at least an indicator to let you know that you have uncommitted changes, but of course it's your call. Great product either way.

**Comment by huyphams** [comment]
Actually, I can add auto-commit to the editor but it can lead to some situation that the user doesn't want to commit.
Adding an option toggle auto-commit is a good idea, but TablePlus can become another tool with complex UI 😄 btw, just let this issue open.

**Comment by vnesek** [comment]
I use tablePlus with MySQL and it is auto-commiting all command I enter in SQL editor which is a bit 
Even if I execute **begin;** and then **update someting** everything it committed right away and there is no way to rollback changes. Buttons for preview/commit/rollback are disabled.
Is there any way to disable auto commit when working in SQL editor.

**Comment by huyphams** [comment]
@vnesek did you modify and press command + S?

**Comment by vnesek** [comment]
I wasn't clear enough. When editing data in table view auto commit is off and I can use cmd+s to commit after making changes. 

Auto commit is always on when manually entering commands in SQL view. It's bad as you can't rollback changes. Safe mode/previewing isn't good enough. Sometimes you have to issue SQL updates and select changes made in transaction to see what changed and then decide if you wan't to commit or rollback. This is immensely useful when making big changes to thousands of rows spanning multiple tables.  Most db clients I've used (SequelPro, Oracle SQLDeveloper, Toad...) support it.

If possible, when user enters **begin;** in SQL view, turn off auto commit, begin transaction and then send commit or rollback either when user executes **commit;** in SQL view or by pressing cmd+s.

**Comment by huyphams** [comment]
I don't think that TablePlus does auto-commit when you `manually entering commands in SQL view.`, basically, TablePlus sends the query from the editor without any modification, and of course, will not append any `COMMIT;`, so how it's possible? I believe that we missed understood something here.

**Comment by huyphams** [comment]
TablePlus was designed for individual and development purpose, so keep the default config of the database and make it simple it the most important thing. If we turn it off by default, there will be a lot of misunderstanding for many users.

I know it would be beneficial for some users but most of them will think it clumsy not necessary.

If you want a transaction, simply run a BEGIN first. If we can improve it, in the future, it would be an option but turn off by default.

**Comment by vnesek** [comment]
Huy, you're doing excellent work with tablePlus. I've bought a license for myself and recommend tablePlus to others. On the other hand, from my personal experience as freelancer over last 20 years, nobody really cares if you're working alone or not after they pay.

Back to the topic, I had this problem yesterday where begin/commit in SQL view just wasn't working. Perhaps I've missed cmd+enter on begin command? Just tried it again and it works as you say (it would be great  that it enables commit button in toolbar after sending begin; to db). If I hit this problem again and can make a test case I'll let you know.

Keep up a good work.

**Comment by huyphams** [comment]
Thanks, @vnesek for the confirmation. I will keep thinking about it. detecting an uncommitted transaction and flag it in the UI is a good idea 👍 I will try to implement it.

**Comment by Krknv** [comment]
One more case. I've trying to test **soft delete** of row. My flow is:
- open tablePlus, finding row _id818_
- open my app, delete row _id818_
- open tablePlus, press **cmd + s** for update row (for cancel deletion)

Actually, for do this - I need to press **cmd + r**, then change some field, then press **cmd + s** for commit.
That's would be great, if I can commit (update) row without any changes (or option for that behavior)

**Comment by huyphams** [comment]
I don't quite understand what your mean @Krknv but you can refresh the current opened table by right click > refresh without any changes.

**Comment by Krknv** [comment]
In other words, I want ability to **press on commit button,** even **nothing's changed** =)

**Comment by huyphams** [comment]
![OK_thumb](https://user-images.githubusercontent.com/3654622/60241166-d1b46b80-98dc-11e9-8bab-b3374d3bb6ec.jpg)

**Comment by gintechsystems** [comment]
There should definitely be a new button that allows auto commit without having to press cmd + s. Since we can customize the toolbar, it can be in there. It doesn't even have to be a default button since you don't seem to want this as a default option, but it should be at least an option. At least for in-line changes on a table, I understand the sql editor portion.

**Comment by humectant** [comment]
Without the ability to roll back commands entered into the SQL editor, this tool cannot be safely used by developers.

**Comment by benzittlau** [comment]
@huyphams I just came across this thread, and I think I understand the conclusions arrived at above, but wanted to re-state them in my own words for clarity. Please correct anything below that doesn't match reality, and hopefully this can help others who find the thread in the future, and avoid misconceptions like:

> Without the ability to roll back commands entered into the SQL editor, this tool cannot be safely used by developers.

Which I don't believe to be accurate or fair based on my analysis of the tool.

## Summary: 
There are two very different ways in which TablePlus will perform SQL against the database to modify data.  
1) By modifying values of rows directly in various views displaying tables of data
2) By using manual SQL directly in the SQL editors

### Modifying values of rows in the table views
In this case, a local modification in TablePlus occurs when the value is modified in a table view, but it appears that no operation is yet performed against the DB itself.  
These changes will be tracked in TablePlus itself, until they are "committed" to the database when you press cmd + s.
"Committing" the changes in this case refers to saving the *local* changes in TablePlus to the DB. There is no pre-existing transaction holding the changes on the DB, until cmd + s is pressed by the user.  When cmd+is *is* pressed, however, a transaction *is* used to apply all the changes that have been queued up, and this transaction is automatically committed when the operations are all successful (and rolled back otherwise).

### Using SQL manually in the SQL editors.
In this case you are fully in control of your own destiny.  TablePlus will merely send what you want to the server.  
If you want a transaction, it's your responsibility to open a transaction yourself by initially sending a "BEGIN" command to the server, and closing it off with a "ROLLBACK" or "COMMIT" command yourself.

## Testing
I believe I've validated those above conclusions in my current version (3.1.0), so everything stated there is "known true".  

The one aspect I'm pretty confident about but a bit less sure than the other aspects is that the local changes in the first method aren't stored in a DB transaction, but are simply locally tracked in TablePlus.  In my testing I was surprised to see that custom SQL select queries run in the SQL editor against the modified row *did* return the unsaved local changes.  I'm guessing this is just TablePlus being clever, as when I made a more complex query utilizing Joins, it no longer returned the unsaved edited value.

**Comment by alvint** [comment]
The worst part about the current behavior is that it's dangerously inconsistent. When connecting to Oracle databases, autocommit is off--you can safely make changes and they won't be persisted unless you type `commit;`. But for other DBs like Postgres, autocommit is on. It's very easy to forget that's the case if you work with multiple DB types. Please reconsider changing this.

**Comment by humectant** [comment]
@benzittlau I agree my statement wasn't accurate and correct and accept the criticism. BEGIN and ROLLBACK can be used to control the transaction; however, when using a tool like TablePlus one expects that this would be built in as a convenience option it is on other SQL editors. This would make it so much more powerful. For potentially destructive work I switch from TablePlus to other editors because I can work so much more fluidly and not have to worry about the transaction markers myself. It would stay within TablePlus if this existed.
